### PR TITLE
Support |DataDirectory| for SQLite provider

### LINKED ISSUE #13840: Microsoft.Data.Sqlite: SQLite provider seems to ignore |DataDirectory|
Moved from https://github.com/aspnet/EntityFrameworkCore/issues/12767 posted by @bdongus

Unlike the SQLserver provider the SQLite provider seems to fail at resolving the placeholder "|DataDirectory|".

```
Microsoft.Data.Sqlite.SqliteException
  HResult=0x80004005
  Nachricht = SQLite Error 14: 'unable to open database file'.
  Quelle = Microsoft.Data.Sqlite
  Stapelüberwachung:
   bei Microsoft.Data.Sqlite.SqliteException.ThrowExceptionForRC(Int32 rc, sqlite3 db)
   bei Microsoft.Data.Sqlite.SqliteConnection.Open()
   bei Microsoft.EntityFrameworkCore.Storage.RelationalConnection.OpenDbConnection(Boolean errorsExpected)
   bei Microsoft.EntityFrameworkCore.Storage.RelationalConnection.Open(Boolean errorsExpected)
   bei Microsoft.EntityFrameworkCore.Sqlite.Storage.Internal.SqliteRelationalConnection.Open(Boolean errorsExpected)
   bei Microsoft.EntityFrameworkCore.Sqlite.Storage.Internal.SqliteDatabaseCreator.Create()
   bei Microsoft.EntityFrameworkCore.Storage.RelationalDatabaseCreator.EnsureCreated()
   bei Microsoft.EntityFrameworkCore.Infrastructure.DatabaseFacade.EnsureCreated()
   bei idee5.Globalization.Test.WithSQLiteBase.MyTestInitalize() in C:\Users\xyz\Source\idee5 intern\idee5.Globalization4netcore\idee5.Globalization.Test\WithSQLiteBase.cs: Zeile23
```

### Steps to reproduce
```c#
var contextOptions = new DbContextOptionsBuilder<GlobalizationDbContext>();
contextOptions.UseSqlite("data source=|DataDirectory|idee5.Resources.db3");
context = new GlobalizationDbContext(contextOptions.Options);
context.Database.EnsureDeleted();
context.Database.EnsureCreated();
```
The exception occurs in EnsureCreated.

### Further technical details
EF Core version: 2.1.1
Database Provider: Microsoft.EntityFrameworkCore.Sqlite 2.1.1
Operating system: Windows 10
IDE: Visual Studio 2017 15.7.5

### PR DESCRIPTION
**Pull request by Muppets**
Summary of the changes
 - Added |DataDirectory| checks in SqliteConnection connection string path
 - Created unit test for checking |DataDirectory| support

Fixes #13840

### CODE REVIEW

File: src/Microsoft.Data.Sqlite.Core/SqliteConnection.cs [hunk 1]
```diff
@@ -215,10 +216,17 @@ public override void Open()
...
             if (!string.IsNullOrEmpty(dataDirectory)
                 && (flags & raw.SQLITE_OPEN_URI) == 0
-                && !filename.Equals(":memory:", StringComparison.OrdinalIgnoreCase)
-                && !Path.IsPathRooted(filename))
+                && !filename.Equals(":memory:", StringComparison.OrdinalIgnoreCase))
             {
-                filename = Path.Combine(dataDirectory, filename);
+                if (filename.StartsWith(DataDirectoryToken, StringComparison.InvariantCultureIgnoreCase))
+                {
+                    filename = Path.Combine(dataDirectory, filename.Substring(DataDirectoryToken.Length));
```
Comment (bricelam) [inline src/Microsoft.Data.Sqlite.Core/SqliteConnection.cs]: EF6 (and I think SqlClient) have some [extra checking](https://github.com/aspnet/EntityFramework6/blob/master/src/EntityFramework/Core/Common/DbProviderServices.cs#L808-L812) to make sure you don't go outside of the data directory. (i.e. `|DataDiretcory|\..\my.db` isn't allowed. Not sure if we want to do the same or not...

File: test/EFCore.Sqlite.FunctionalTests/SqliteDatabaseCreatorTest.cs [hunk 2]
```diff
@@ -57,6 +59,39 @@ public async Task Exists_returns_true_when_database_exists(bool async, bool useC
             }
         }
 
+        [ConditionalTheory]
+        [InlineData(false, false)]
+        [InlineData(true, false)]
+        [InlineData(false, true)]
+        [InlineData(true, true)]
+        public async Task Create_database_with_custom_DataDirectory_path(bool async, bool useCanConnect)
```
Comment (bricelam) [inline test/EFCore.Sqlite.FunctionalTests/SqliteDatabaseCreatorTest.cs]: 😄 Not sure we need an EF Core test for this. Can you add a test to `Microsoft.Data.Sqlite.Tests` instead?

File: src/Microsoft.Data.Sqlite.Core/SqliteConnection.cs (hunk 1 above)
Comment (bricelam) [reply src/Microsoft.Data.Sqlite.Core/SqliteConnection.cs]: I suppose we don't block it in relative paths, so we probably shouldn't with the macro.

File: test/EFCore.Sqlite.FunctionalTests/SqliteDatabaseCreatorTest.cs (hunk 2 above)
Comment (Muppets) [reply test/EFCore.Sqlite.FunctionalTests/SqliteDatabaseCreatorTest.cs]: Yeah nice, made for a much nicer test too.

**Comment by bricelam** [comment]
I wonder if `sqliteConnection.DataSource` should also expand this when the connection is closed...

**Comment by bricelam** [comment]
(Lets see what SQL CE does...)

**Comment by bricelam** [comment]
Looks like CE just returns the string as it was specified in the connection string (always). I'm fine having the same behavior for unopened connections.

We already have a similar difference with relative paths (they're made absolute when open).

File: src/Microsoft.Data.Sqlite.Core/SqliteConnection.cs [hunk 3]
```diff
@@ -19,6 +19,7 @@ namespace Microsoft.Data.Sqlite
     public partial class SqliteConnection : DbConnection
     {
         internal const string MainDatabaseName = "main";
+        internal const string DataDirectoryToken = "|DataDirectory|";
```
Comment (bricelam) [inline src/Microsoft.Data.Sqlite.Core/SqliteConnection.cs]: Can be `private`.

Comment (bricelam) [reply src/Microsoft.Data.Sqlite.Core/SqliteConnection.cs]: Nit: I'd call this `DataDirectoryMacro`, but I'm also very jaded.

File: test/Microsoft.Data.Sqlite.Tests/SqliteConnectionTest.cs [hunk 4]
```diff
@@ -127,9 +127,37 @@ public void Open_throws_when_no_connection_string()
...
+            {
+                AppDomain.CurrentDomain.SetData("DataDirectory", null);
+            }
+        }
+
         [Fact]
         public void Open_adjusts_relative_path()
         {
+            // Ensure there isn't a data directory context that can affect the pathing
+            AppDomain.CurrentDomain.SetData("DataDirectory", null);
```
Comment (bricelam) [inline test/Microsoft.Data.Sqlite.Tests/SqliteConnectionTest.cs]: Nit: Not really needed

**Comment by ErikEJ** [comment]
var connStr = "Data Source=|DataDirectory|\\test.sdf";
            var engine = new SqlCeEngine(connStr);
            engine.CreateDatabase();

            var conn = new SqlCeConnection(connStr);
            conn.Open();
            Console.WriteLine(conn.DataSource);

Returns: 

       "Data Source=|DataDirectory|\test.sdf"

**Comment by bricelam** [comment]
Thanks, @ErikEJ! I was just skimming through source code, so i'm glad you could confirm the behavior.

**Comment by NeilBostrom** [comment]
@bricelam I'd made those review changes, all seemed good to me. Nit and all!

**Pull request merged**